File: src/main/java/com/kingdom/util/specialaction/LeaderSpecialActionHandler.java
package com.kingdom.util.specialaction;

import com.kingdom.model.Card;
import com.kingdom.model.CardAction;
import com.kingdom.model.Game;
import com.kingdom.model.Player;

/**
 * Created by IntelliJ IDEA.
 * Date: 8/29/11
 * Time: 8:09 PM
 */
public class LeaderSpecialActionHandler {

    public static void handleSpecialAction(Game game, Card card) {
        Player player = game.getCurrentPlayer();

        if (card.getName().equals("Archimedes")) {
            player.setBuyBonusTurns(2);
            player.setEnableVictoryCardDiscount(true);
        } else if (card.getName().equals("Aristotle")) {
            player.setCardBonusTurns(2);
        } else if (card.getName().equals("Bilkis")) {
            CardAction cardAction = new CardAction(CardAction.TYPE_CHOOSE_CARDS);
            cardAction.setDeck(Card.DECK_LEADERS);
            cardAction.setCardName(card.getName());
            cardAction.setButtonValue("Done");
            cardAction.setNumCards(1);
            cardAction.setInstructions("Select one of the following cards to gain on top of your deck and then click Done.");
            for (Card c : game.getSupplyMap().values()) {
                if (game.getCardCost(c) <= 6 && !c.isCostIncludesPotion() && game.getSupply().get(c.getCardId()) > 0) {
                    cardAction.getCards().add(c);
                }
            }
            if (cardAction.getCards().size() > 0) {
                game.setPlayerCardAction(player, cardAction);
            }
        } else if (card.getName().equals("Imhotep")) {
            player.setBuyBonusTurns(2);
            player.setEnableTreasureCardDiscount(true);
        } else if (card.getName().equals("Leonidas")) {
            player.setBuyBonusTurns(2);
            player.setEnableActionCardDiscount(true);
        } else if (card.getName().equals("Maecenas")) {
            player.setLeaderDiscount(3);
        } else if (card.getName().equals("Plato")) {
            if (player.getHand().size() > 0) {
                CardAction cardAction = new CardAction(CardAction.TYPE_TRASH_UP_TO_FROM_HAND);
                cardAction.setDeck(Card.DECK_LEADERS);
                cardAction.setCardName(card.getName());
                cardAction.setButtonValue("Done");
                cardAction.setNumCards(2);
                cardAction.setInstructions("Trash up to 2 cards.");
                cardAction.setCards(player.getHand());
                game.setPlayerCardAction(player, cardAction);
            }
        } else if (card.getName().equals("Tomyris")) {
            int nextPlayerIndex = game.calculateNextPlayerIndex(game.getCurrentPlayerIndex());
            while (nextPlayerIndex != game.getCurrentPlayerIndex()) {
                Player nextPlayer = game.getPlayers().get(nextPlayerIndex);
                int numInSupply = game.getSupply().get(Card.CURSE_ID);
                if (numInSupply > 0) {
                    game.playerGainedCard(nextPlayer, game.getCurseCard());
                    game.refreshDiscard(nextPlayer);
                }
                nextPlayerIndex = game.calculateNextPlayerIndex(nextPlayerIndex);
            }
        } else if (card.getName().equals("Xenophon")) {
            player.addCoins(2);
            player.addBuys(1);
            game.refreshAllPlayersPlayingArea();
        }
    }
}

File: src/main/kotlin/com/kingdom/util/specialaction/LeaderSpecialActionHandler.kt
package com.kingdom.util.specialaction

import com.kingdom.model.*
import com.kingdom.model.cards.Card
import com.kingdom.model.cards.Deck
import com.kingdom.model.cards.supply.Curse

object LeaderSpecialActionHandler {

    fun handleSpecialAction(game: Game, card: Card) {
        val player = game.currentPlayer

        when (card.name) {
            "Archimedes" -> {
                player!!.buyBonusTurns = 2
                player.setEnableVictoryCardDiscount(true)
            }
            "Aristotle" -> player!!.cardBonusTurns = 2
            "Bilkis" -> {
                val cardAction = OldCardAction(OldCardAction.TYPE_CHOOSE_CARDS)
                cardAction.deck = Deck.Leaders
                cardAction.cardName = card.name
                cardAction.buttonValue = "Done"
                cardAction.numCards = 1
                cardAction.instructions = "Select one of the following cards to gain on top of your deck and then click Done."
                for (c in game.supplyMap.values) {
                    if (game.getCardCost(c) <= 6 && !c.costIncludesPotion && game.supply[c.name]!! > 0) {
                        cardAction.cards.add(c)
                    }
                }
                if (cardAction.cards.size > 0) {
                    game.setPlayerCardAction(player!!, cardAction)
                }
            }
            "Imhotep" -> {
                player!!.buyBonusTurns = 2
                player.setEnableTreasureCardDiscount(true)
            }
            "Leonidas" -> {
                player!!.buyBonusTurns = 2
                player.setEnableActionCardDiscount(true)
            }
            "Maecenas" -> player!!.setLeaderDiscount(3)
            "Plato" -> if (player!!.hand.size > 0) {
                val cardAction = OldCardAction(OldCardAction.TYPE_TRASH_UP_TO_FROM_HAND)
                cardAction.deck = Deck.Leaders
                cardAction.cardName = card.name
                cardAction.buttonValue = "Done"
                cardAction.numCards = 2
                cardAction.instructions = "Trash up to 2 cards."
                cardAction.cards = player.hand
                game.setPlayerCardAction(player, cardAction)
            }
            "Tomyris" -> {
                var nextPlayerIndex = game.calculateNextPlayerIndex(game.currentPlayerIndex)
                while (nextPlayerIndex != game.currentPlayerIndex) {
                    val nextPlayer = game.players[nextPlayerIndex]
                    val numInSupply = game.supply[Curse.NAME]!!
                    if (numInSupply > 0) {
                        game.playerGainedCard(nextPlayer, game.curseCard)
                        game.refreshDiscard(nextPlayer)
                    }
                    nextPlayerIndex = game.calculateNextPlayerIndex(nextPlayerIndex)
                }
            }
            "Xenophon" -> {
                player!!.addCoins(2)
                player.addBuys(1)
                game.refreshAllPlayersPlayingArea()
            }
        }
    }
}
